API-Specification.md

```diff
@@ -1,19 +1,264 @@
+Content-Length: 31
+
+{"message":"Cannot parse JSON"}
+```
+
+2. **Incorrect JSON value types**: The server MUST return with `400 Bad Request`
+response.
+
+```http
+HTTP/1.1 400 Bad Request
+Content-Length: 40
+
+{"message":"Incorrect JSON value types"}
+```
+
+3. **Invalid fields**: The server MUST return with `422 Unprocessable Entity`
+response.
+
+```http
+HTTP/1.1 422 Unprocessable Entity
+Content-Length: 149
+
+{
+  "message": "Validation Failed",
+  "errors": [
+    {
+      "resource": "Library",
+      "field": "title",
+      "code": "missing-field"
+    }
+  ]
+}
+```
+
+Each `error` object in `errors` SHOULD contain `resource`, `field` and `code`
+entries for diagnostics.
+
+Possible validation error codes include:
+
+| Error Code        | Description                                   |
+| ----------------- | --------------------------------------------- |
+| `missing`         | Resource does not exist.                      |
+| `missing-field`   | Required field missing.                       |
+| `invalid`         | Incorrect formatting.                         |
+| `duplicate`       | Resource with identical value already exists. |
+
+Resources MAY also respond with `custom` error codes, which means they MUST
+then have a complementary `message` field for describing the problem.
+
+For example,
+
+```http
+HTTP/1.1 422 Unprocessable Entity
+Content-Length: 191
+
+{
+  "message": "Validation Failed",
+  "errors": [
+    {
+      "resource": "Library",
+      "field": "title",
+      "code": "custom",
+      "message": "UCL API service unavailable"
+    }
+  ]
+}
+```
+
+## HTTP Redirects
+
+Clients SHOULD assume that any requests to resources may cause a redirect
+response, which clients SHOULD follow.
+
+Redirect responses MUST contain the `Location` header field which specifies the
+destination URI of the resource.
+
+| HTTP Status Code | Description                                               |
+| ---------------- | --------------------------------------------------------- |
+| `301`            | Permanent redirection. Clients SHOULD migrate to new URI. |
+| `302`, `307`     | Temporary redirection. Clients SHOULD use original URI.   |
+
+## HTTP Verbs
+
+Suitable HTTP Verbs SHOULD be used to convey action semantics:
+
+| HTTP Verb | Description                   |
+| --------- | ----------------------------- |
+| `HEAD`    | Get only header information.  |
+| `GET`     | Fetching resource.            |
+| `POST`    | Create resource.              |
+| `PATCH`   | Update resource.              |
+| `PUT`     | Replace resource.             |
+| `DELETE`  | Delete resource.              |
+
+## Pagination
+
+Requests which return list containing multiple items SHOULD be paginated with
+30 items be default. Clients SHOULD be able to specify custom pages via
+the `?page={PAGE_NUMBER}` query parameter. Page numbers MUST begin at `1`.
+Clients SHOULD be able to specify items per page via `?per_page={NUM_ITEMS}`.
```

Comment:
This would be a nice way, but for post feeds' pagination, currently, the method is by specifying a starting postId from client, eg `?startid=203`, which is the last post of the first page. And the server will respond next 12 posts starting from id 204. This will then concat to frontend. Using this way it would prevent the case that if there are more posts added, and page 2 is no longer holding original posts the client wanted

Reply:
This is general pagination. We can and should add an additional section for feed pagination. Alternatively, can this be handled by caching last-viewed post IDs from the client side?

So maybe something like `/feed/?offset=204&per_page=12`, and page number can be calculated by `204 div 12 == 17`.

Reply:
In fact the client should probably be doing this anyway. The backend is not going to store any client side state.

Reply:
Yes, that would do, and the page number is not even necessary in that case